### PR TITLE
Added a script to perform " Validation of constant flushing of data when SSD write back cache feature is enabled"

### PR DESCRIPTION
**Pull request by preethisn**
Signed-off-by: Preethi Nataraj <pnataraj@localhost.localdomain>

# Description validate constant flushing and evicting of data for an extended period of time.

### CODE REVIEW

**Comment by preethisn** [comment]
logs copied to http://magna002.ceph.redhat.com/ceph-qe-logs/Preethi/ssd.txt

File: utility/SSD_writeback_cache.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    This method to validate constant flushing and
+    evicting of data for an extended period of time.
+    This scenario in particular
+    recommend writing a script that would run through a few thousand
+    iterations of SIGKILLed "rbd bench" "fio" on the same image:
+
+    """
+    size = 1 * 1024 * 3
+    pool_name = "ssd"
+    image_name = "image3"
```
Comment (psathyan) [inline utility/SSD_writeback_cache.py:30]: Would it make sense to retrieve these values from CLI? 

The current implementation creates the required resources but does not cleanup after the workflow completion. Can we implement a teardown sequence?

File: utility/SSD_writeback_cache.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    os.system(f"ceph osd pool create {pool_name} 128")
+    os.system(f"ceph osd pool application enable {pool_name} rbd")
+    os.system(f"rbd pool init -p {pool_name}")
+    os.system(f"rbd create --size {size} {pool_name}/{image_name} --thick-provision")
```
Comment (psathyan) [inline utility/SSD_writeback_cache.py:34]: These resources are required for testing... please verify that they are created successfully before proceeding to the next step.

File: utility/SSD_writeback_cache.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    for i in range(1, 10000):
+        print("#" * 80)
```
Comment (psathyan) [inline utility/SSD_writeback_cache.py:36]: IMO, we can avoid this line.... we would have not many

File: utility/SSD_writeback_cache.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+    """
+    size = 1 * 1024 * 3
+    pool_name = "ssd"
+    image_name = "image3"
+    os.system(f"ceph osd pool create {pool_name} 128")
+    os.system(f"ceph osd pool application enable {pool_name} rbd")
+    os.system(f"rbd pool init -p {pool_name}")
+    os.system(f"rbd create --size {size} {pool_name}/{image_name} --thick-provision")
+    for i in range(1, 10000):
```
Comment (psathyan) [inline utility/SSD_writeback_cache.py:35]: Can this be made as a configurable?

File: utility/SSD_writeback_cache.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        proc = subprocess.Popen(
+            [
+                f"fio --name=test-1 --ioengine=rbd --pool={pool_name} "
+                f"--rbdname={image_name} --numjobs=2 --rw=write --bs=4k "
+                f"--iodepth=8 --fsync=32 --runtime=80 --time_based "
+                f"--group_reporting --ramp_time=120"
+            ],
+            shell=True,
+        )
+        timeout = random.randint(45, 100)
```
Comment (psathyan) [inline utility/SSD_writeback_cache.py:48]: IMO, we should check the storage capacity before going to the next iteration. We could break the cluster if it is not properly sized.

File: utility/SSD_writeback_cache.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        print(f"sleeping for {timeout}")
```
Comment (psathyan) [inline utility/SSD_writeback_cache.py:49]: opt to use `log` instead of `print`

File: utility/SSD_writeback_cache.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        log.info("image status %s" % output)
+        image_info = json.loads(output)
+        if image_info.get("image_cache_state"):
+            cache_state = json.loads(image_info["image_cache_state"])
+            if cache_state.get("clean") == "false":
+                raise AssertionError("image status is not in clean state")
+
+
+if __name__ == "__main__":
+    rc = ssd_write_cache()
```
Comment (psathyan) [inline utility/SSD_writeback_cache.py:64]: There is no return value for the method `ssd_write_cache`

File: utility/SSD_writeback_cache.py (hunk 1 above)
Comment (sunilkumarn417) [reply utility/SSD_writeback_cache.py:30]: https://github.com/red-hat-storage/cephci/blob/master/pipeline/scripts/ci/cos_cli.py please refer this for custom values which should be picked from CLI.

Comment (preethisn) [reply utility/SSD_writeback_cache.py]: Yes, implemented via cli with custom values as per the reference provided

File: utility/SSD_writeback_cache.py (hunk 5 above)
Comment (preethisn) [reply utility/SSD_writeback_cache.py]: As discussed, will consider this storage capacity as improvements and enhancement in next sprint

File: utility/SSD_writeback_cache.py (hunk 6 above)
Comment (preethisn) [reply utility/SSD_writeback_cache.py]: done

**Comment by preethisn** [comment]
New logs attached 
[logs_ssd.txt](https://github.com/red-hat-storage/cephci/files/8178998/logs_ssd.txt)

File: utility/SSD_writeback_cache.py (hunk 2 above)
Comment (preethisn) [reply utility/SSD_writeback_cache.py]: Yes, these are working fine and captured in logging.

**Review by psathyan — COMMENTED** [review]
minor changes...

**Comment by preethisn** [comment]
Changes are done. Please find the latest logs.
[logs_ssd.txt](https://github.com/red-hat-storage/cephci/files/8184474/logs_ssd.txt)

File: utility/ssd_writeback_cache.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    level=logging.INFO,
+    format="%(asctime)s - %(levelname)s - %(filename)s:%(lineno)d - %(message)s",
+)
+
+doc = """
+
+This script helps to validate the write back cache flushing of data
+
+usage:
+SSD_writeback_cache.py  --pool POOL --image IMAGE [--size=<size>] [--iteration=<iteration>]
```
Comment (sunilkumarn417) [inline utility/ssd_writeback_cache.py:24]: please replace module name here.

File: utility/ssd_writeback_cache.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+Options:
+  -h --help                 Show this screen
+  --pool <name>             Name of the pool
+  --image <name>            Image name
+  --size=<size>             Size of the image default is 1G     [default: 1]
+  --iteration=<iteration>   No of iterations for Ios to run     [default: 10000]
+"""
+
+
+def ssd_write_cache(args):
```
Comment (sunilkumarn417) [inline utility/ssd_writeback_cache.py:36]: I think we might not need parameter `args`.

File: utility/ssd_writeback_cache.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    )
+
+
+if __name__ == "__main__":
+    args = docopt(doc)
+    image_size = 1024 * int(args["--size"])
+    iteration = int(args["--iteration"])
+    pool_name = args["--pool"]
+    image_name = args["--image"]
+    ssd_write_cache(args)
```
Comment (sunilkumarn417) [inline utility/ssd_writeback_cache.py:94]: please include try-finally clause to have automatic teardown of RBD pool.

Comment (preethisn) [reply utility/ssd_writeback_cache.py:94]: Thanks Sunil. We take this input in our next PR

File: utility/SSD_writeback_cache.py (hunk 4 above)
Comment (preethisn) [reply utility/SSD_writeback_cache.py]: i dont think so we can make it as configurable

File: utility/ssd_writeback_cache.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+    os.system(
+        f"ceph osd pool delete {pool_name} {pool_name} --yes-i-really-really-mean-it "
+    )
+
+
+if __name__ == "__main__":
+    args = docopt(doc)
+    image_size = 1024
+    if args.get("--size"):
+        image_size = int(args["<size>"]) * image_size
```
[the remark this answers is not in the document]
Comment (preethisn) [reply utility/ssd_writeback_cache.py]: This change is implemented as suggested.

File: utility/ssd_writeback_cache.py (hunk 8 above)
Comment (preethisn) [reply utility/ssd_writeback_cache.py]: done

File: utility/ssd_writeback_cache.py (hunk 9 above)
Comment (preethisn) [reply utility/ssd_writeback_cache.py]: removed and tested. its working fine

**Comment by preethisn** [comment]
@sunilkumarn417 @psathyan I have addressed the review comments except below and we consider them in our next PR. 
- please include try-finally clause to have automatic teardown of RBD pool.
- calculate storage capacity before moving to next iteration
Please take this PR to closure if no concerns.

**Review by psathyan — COMMENTED** [review]
This looks good as the plan is to take the improvements later

**Review by sunilkumarn417 — APPROVED** [review]
As improvements is considered in next PR, let's get this code in.

**Pull request merged**